### PR TITLE
faster and more readable array transfer function

### PR DESCRIPTION
**Pull request by MMesch**
This little modification is makes the computation of the array transfer function more readable, and it is 6x faster. It uses more memory though but 2D grids are usually still ok in memory. Not sure which one you prefer. ~18ms vs 1.33s are the run times of the array example.

### CODE REVIEW

File: obspy/signal/array_analysis.py [hunk 1]
```diff
@@ -731,7 +731,6 @@ def get_spoint(stream, stime, etime):
                         tr.stats.sampling_rate + .5)
     return spoint, epoint
 
-
```
Comment (QuLogic) [inline obspy/signal/array_analysis.py]: This line should not be removed.

File: obspy/signal/array_analysis.py [hunk 2]
```diff
@@ -834,7 +830,6 @@ def array_transff_freqslowness(coords, slim, sstep, fmin, fmax, fstep,
     transff /= transff.max()
     return transff
 
-
```
Comment (QuLogic) [inline obspy/signal/array_analysis.py]: This line should also not be removed.

File: obspy/signal/array_analysis.py [hunk 3]
```diff
@@ -763,18 +762,15 @@ def array_transff_wavenumber(coords, klim, kstep, coordsys='lonlat'):
...
-            for k in range(len(coords)):
-                _sum += np.exp(complex(0.,
-                               coords[k, 0] * kx + coords[k, 1] * ky))
-            transff[i, j] = abs(_sum) ** 2
+    ks  = np.transpose(np.vstack( (kxgrid.flatten(),kygrid.flatten()) ))
 
-    transff /= transff.max()
-    return transff
+    k_dot_r = np.einsum('ni,mi->nm',ks,coords[:,:2]) #z coordinate is not used
+    transff = np.abs(np.sum(np.exp(1j*k_dot_r), axis=1))**2/len(coords)**2
```
Comment (QuLogic) [inline obspy/signal/array_analysis.py]: Please use PEP8 style; a space is required around operators and after commas and comment characters.

File: obspy/signal/array_analysis.py [hunk 4]
```diff
@@ -763,18 +762,15 @@ def array_transff_wavenumber(coords, klim, kstep, coordsys='lonlat'):
...
+
+    return transff.reshape(nkx,nky)
```
Comment (QuLogic) [inline obspy/signal/array_analysis.py]: I'm pretty sure a set of parentheses is missing here.

**Comment by QuLogic** [comment]
Unfortunately, the results are not quite the same; it is fine if there are small fluctuations due to differing orders of operations, but these are 20-70% of the original values.

**Comment by MMesch** [comment]
Oups, I'll have a look into this! It should be the same. Probably because there is no normalization in the end?

**Comment by MMesch** [comment]
ok I think there is a turned axis.

**Comment by MMesch** [comment]
OK. I confirm the result is exactly the transpose of the previous function.

**Comment by MMesch** [comment]
OK. Its fine now. I think in general it would be better to make a function that takes wavenumbers and positions as input instead of klim and kstep and returns the transfer function for each k. This leads to better organisation of the code and makes the function more flexible. What do you think about this?

**Comment by megies** [comment]
> A further advantage is that it automatically works for 2D ks + 2D position or 3D ks and 3D positions. What do you think about this?

I don't use it a lot but other people might well do. Maybe it would be an option to keep the current routine and make it a wrapper around a new function with your recommended call parameters (where most of the current code would then reside, I guess)?

**Comment by krischer** [comment]
This can be merged, right?

**Comment by MMesch** [comment]
yes from my side...

**Pull request merged**